src/engine/enginemaster.cpp

```diff
@@ -366,16 +406,80 @@ void EngineMaster::process(const int iBufferSize) {
     // Mix all the PFL enabled channels together.
     m_headphoneGain.setGain(chead_gain);
 
-    if (m_bRampingGain) {
-        ChannelMixer::mixChannelsRamping(
-                m_headphoneGain, &m_activeHeadphoneChannels,
-                &m_channelHeadphoneGainCache,
-                m_pHead, iBufferSize);
+    // Mixxx gets input signals delayed by the input latency,
+    // so to preserve the relative timing of the input with the master and headphones
+    // outputs, delay the master, booth, and headphones outputs by the input latency.
+    // This way, if the microphone user is on beat with the output they hear coming from
+    // the speakers/headphones, the input will remain on beat at the expense of
+    // adding some latency.
+    // If the input signal is being directly monitored (mixed with the master output
+    // in hardware), do not delay the master output. The recording/broadcasting signal
+    // will be delayed in this case to keep the recorded/broadcasted signal on beat
+    // and consistent with what the user hears out the speakers/headphones.
```

Comment:
With the direct monitor option, the user would configure their sound card to mix the microphone with the main output in hardware, configure the mic input in Mixxx, and press the talk button on the mic input. In this case, Mixxx does not mix the mic input with the master output because that is already done in hardware, but it is necessary to mix the mic input into the record/broadcast signal.